### PR TITLE
Shahar app

### LINKED ISSUE #13: Dashborad page
dashboard multiple entries page.
dashboard single entry view.

### LINKED ISSUE #14: Thief Page
thief multiple entries page.
thief single entry view.

### LINKED ISSUE #15: Thief bloc Tests

### PR DESCRIPTION
**Pull request by shaharewq0**
closes #13 
closes #14 
closes #15
dealt with warnings

### CODE REVIEW

**Comment by TomerMe2** [comment]
Can you put pictures of the pages that you've created?

File: security_drone_user_app/lib/logic/thief_page_bloc.dart [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  ThiefPageBloc.test(ThiefEntry entry) : super(ShowThiefEntries([entry]));
+
+  @override
+  Stream<ThiefPageState> mapEventToState(ThiefPageEvent event) async* {
+
+    if (event is ThiefEntryClicked){
+      yield ShowThiefEntry(state.entries, event.clickedOn);
+    }
+    else if (event is RefreshThiefEntries) {
+      print('api request to refresh the entries');
```
Comment (TomerMe2) [inline security_drone_user_app/lib/logic/thief_page_bloc.dart:25]: Remove the print.

File: security_drone_user_app/lib/logic/thief_page_bloc.dart [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      List<ThiefEntry> entries = ThiefEntry.dummyFetchAll();
+      yield ShowThiefEntries(entries);
+    }
+    else if (event is DisplayThiefEntries) {
+      yield ShowThiefEntries(state.entries);
+    }
+    else {
+      throw Exception("Event not found $event");
```
Comment (TomerMe2) [inline security_drone_user_app/lib/logic/thief_page_bloc.dart:33]: Remove the exception.

File: security_drone_user_app/lib/main.dart [hunk 3]
```diff
@@ -50,25 +58,28 @@ class MyHomePage extends StatefulWidget {
...
         title: Text(widget.title),
+        actions: [
+          ImageBanner('assets/images/icon.jpg')
+        ],
+      ),
+      body: Column(
+        mainAxisAlignment: MainAxisAlignment.center,
+        crossAxisAlignment: CrossAxisAlignment.start,
+        children: [
+          TextSection("HELLOOOOO", "hellohellohellohellohellohellohellohellohellohello")
```
Comment (TomerMe2) [inline security_drone_user_app/lib/main.dart:76]: Remove this

File: security_drone_user_app/lib/presentation/image_banner.dart [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+  ImageBanner(this._path);
+
+  @override
+  Widget build(BuildContext context) {
+    return  Row(
+        mainAxisAlignment: MainAxisAlignment.start,
+      children: [Container(
+      constraints: BoxConstraints.expand(
+        height: 100.0,
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/image_banner.dart:14]: Make height and width as parameters of the constructor of the class, you can default them to 100.

File: security_drone_user_app/lib/presentation/pages/dashboard_page.dart [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  State<StatefulWidget> createState() {
+    return DashboardPageState();
+  }
+}
+
+class DashboardPageState extends State<DashboardPage>{
+
+  @override
+  Widget build(BuildContext context){
+    // ignore: close_sinks
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/dashboard_page.dart:22]: Please handle this warning

File: security_drone_user_app/lib/presentation/pages/dashboard_page.dart [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                RefreshDashboardEntries(),
+              ),
+            ),
+          ],
+        )
+    );
+  }
+}
+
+// ignore: must_be_immutable
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/dashboard_page.dart:42]: Please handle this warning

File: security_drone_user_app/lib/presentation/pages/dashboard_page.dart [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+  @override
+  Widget build(BuildContext context){
+    return SingleChildScrollView(
+      padding: EdgeInsets.all(10.0),
+      child: BlocConsumer<DashboardBloc, DashboardState>(
+        cubit: _bloc,
+        buildWhen: (prev, curr){
+          return curr.entries.isNotEmpty;
+        },
+        // ignore: missing_return,
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/dashboard_page.dart:77]: Please handle this warning

File: security_drone_user_app/lib/presentation/pages/dashboard_page.dart [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        builder: (context, state) {
+          if (state is ShowDashboardEntries) {
+            var builder = ListView.builder(
+                physics: ScrollPhysics(),
+                shrinkWrap: true,
+                padding: EdgeInsets.all(10.0),
+                itemCount: state.entries.length,
+                itemBuilder: (context, index) {
+                  var color;
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/dashboard_page.dart:86]: Problematic to write var here, because you don't initialize it. It means that the type will be dynamic, you should avoid dynamic as much as you can. Write Color color or something like that. Reference to the var thingy: https://stackoverflow.com/questions/12416507/difference-between-var-and-dynamic-type-in-dart

File: security_drone_user_app/lib/presentation/pages/dashboard_page.dart [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+            );
+            return builder;
+          }
+          else if (state is ShowDashboardEntry){
+            var index = state.focusedIndex;
+            var display = Column(
+              children: [
+                TextSection(
+                    "Action Full Desctiption:",
+                    state.entries[index].toString()
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/dashboard_page.dart:118]: Not a good practice. I guess that each data entry from this entry (date, mission type, coordinates ...) should have a different style. You can't do that using toString.

File: security_drone_user_app/lib/presentation/pages/dashboard_page.dart [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+              ],
+            );
+            return display;
+          }
+          else {
+            return Text("OK");
+          }
+        },
+        listener: (BuildContext context, state) {
+          /*
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/dashboard_page.dart:137]: What should be here? Remove the comments.

File: security_drone_user_app/lib/presentation/pages/dashboard_page.dart [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+                    ),
+                    onPressed: () => _bloc.add(
+                        DisplayDashboardEntries()
+                    ))
+              ],
+            );
+            return display;
+          }
+          else {
+            return Text("OK");
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/dashboard_page.dart:133]: What this return means?

File: security_drone_user_app/lib/presentation/pages/patrol_set_page.dart [hunk 12]
```diff
@@ -32,16 +30,21 @@ class PatrolSetPageState extends State<PatrolSetPage> {
         padding: EdgeInsets.all(3),
         child: SizedBox(
           width: 75,
-          child: FlatButton(
-            padding: EdgeInsets.all(0),
+          child: TextButton(
+            //padding: EdgeInsets.all(0),
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/patrol_set_page.dart:34]: Why did you change FlatButton to TextButton and removed the 0 padding?

File: security_drone_user_app/lib/presentation/pages/patrol_set_page.dart [hunk 13]
```diff
@@ -32,16 +30,21 @@ class PatrolSetPageState extends State<PatrolSetPage> {
...
             onPressed: func,
             child: Text(text,
                 overflow: TextOverflow.visible,
                 textAlign: TextAlign.center,
                 style: TextStyle(fontSize: 13)
             ),
-            color: Colors.transparent,
-            highlightColor: Colors.blueAccent,
+            style: ButtonStyle(
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/patrol_set_page.dart:41]: Continue the previous comment

File: security_drone_user_app/lib/presentation/pages/patrol_set_page.dart [hunk 14]
```diff
@@ -85,20 +88,25 @@ class PatrolSetPageState extends State<PatrolSetPage> {
...
                   PatrolMapPointClicked(LatLngPoint(loc.latitude, loc.longitude)));
             },
+
           ),
           Align(
             alignment: Alignment.topRight,
             child: Padding(
               padding: EdgeInsets.only(right: 5),
-              child: FlatButton(
+              child: TextButton(
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/patrol_set_page.dart:97]: The previous comment

File: security_drone_user_app/lib/presentation/pages/view_thief_page.dart [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+  State<StatefulWidget> createState() {
+    return ViewThiefPageState();
+  }
+}
+
+class ViewThiefPageState extends State<ViewThiefPage>{
+
+  @override
+  Widget build(BuildContext context){
+    // ignore: close_sinks
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/view_thief_page.dart:20]: Handle the warning

File: security_drone_user_app/lib/presentation/pages/view_thief_page.dart [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+              RefreshThiefEntries(),
+            ),
+          ),
+        ],
+      )
+    );
+  }
+}
+
+// ignore: must_be_immutable
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/view_thief_page.dart:40]: Handle the warning

File: security_drone_user_app/lib/presentation/pages/view_thief_page.dart [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+class ThiefListClass extends StatelessWidget {
+  ThiefPageBloc _bloc;
+
+  ThiefListClass(this._bloc);
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/view_thief_page.dart:48]: The sacffold with the appbar should be in the main of the application. This should not be implemented in every widget.

File: security_drone_user_app/lib/presentation/pages/view_thief_page.dart [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+        actions: [
+          ImageBanner('assets/images/icon.jpg')
+        ],
+      ),
+      body: ThiefList(_bloc),
+    );
+  }
+}
+
+// ignore: must_be_immutable
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/view_thief_page.dart:60]: Handle the warning.

File: security_drone_user_app/lib/presentation/pages/view_thief_page.dart [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+  @override
+  Widget build(BuildContext context){
+    return SingleChildScrollView(
+      padding: EdgeInsets.all(10.0),
+      child: BlocConsumer<ThiefPageBloc, ThiefPageState>(
+        cubit: _bloc,
+        buildWhen: (prev, curr){
+          return curr.entries.isNotEmpty;
+        },
+        // ignore: missing_return,
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/view_thief_page.dart:75]: Handle the warning.

File: security_drone_user_app/lib/presentation/pages/view_thief_page.dart [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+
+  ThiefList(this._bloc);
+
+  @override
+  Widget build(BuildContext context){
+    return SingleChildScrollView(
+      padding: EdgeInsets.all(10.0),
+      child: BlocConsumer<ThiefPageBloc, ThiefPageState>(
+        cubit: _bloc,
+        buildWhen: (prev, curr){
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/view_thief_page.dart:72]: Space before {

File: security_drone_user_app/lib/presentation/pages/view_thief_page.dart [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+// ignore: must_be_immutable
+class ThiefList extends StatelessWidget {
+  ThiefPageBloc _bloc;
+
+  ThiefList(this._bloc);
+
+  @override
+  Widget build(BuildContext context){
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/view_thief_page.dart:67]: Space before {

File: security_drone_user_app/lib/presentation/pages/view_thief_page.dart [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+                physics: ScrollPhysics(),
+                shrinkWrap: true,
+                padding: EdgeInsets.all(10.0),
+                itemCount: state.entries.length,
+                itemBuilder: (context, index) {
+                  return Card(
+                    child: ListTile(
+                      contentPadding: EdgeInsets.symmetric(horizontal: 50),
+                      title: state.entries[index].image,
+                      subtitle: Text(index.toString()),
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/view_thief_page.dart:88]: Same comment on toString

File: security_drone_user_app/lib/presentation/pages/view_thief_page.dart [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+            );
+            return builder;
+          }
+          else if (state is ShowThiefEntry){
+            var index = state.focusedIndex;
+            var display = Column(
+              children: [
+                state.entries[index].image,
+                TextSection(
+                  state.entries[index].date.toString(),
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/view_thief_page.dart:102]: Don't use toString on date, use formatting that you'd like to show.
Reference: https://stackoverflow.com/questions/16126579/how-do-i-format-a-date-with-dart

File: security_drone_user_app/lib/presentation/pages/view_thief_page.dart [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+              ],
+            );
+            return display;
+          }
+          else {
+            return Text("OK");
+          }
+        },
+        listener: (BuildContext context, state) {
+          /*
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/view_thief_page.dart:122]: What should be here?
Remove the comment

File: security_drone_user_app/lib/presentation/pages/view_thief_page.dart [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+                    ),
+                    onPressed: () => _bloc.add(
+                      DisplayThiefEntries()
+                    ))
+              ],
+            );
+            return display;
+          }
+          else {
+            return Text("OK");
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/pages/view_thief_page.dart:118]: What does this OK means?

File: security_drone_user_app/lib/presentation/text_section.dart [hunk 26]
```diff
@@ -0,0 +1,28 @@
+import 'package:flutter/material.dart';
+import 'package:security_drone_user_app/style.dart';
+
+class TextSection extends StatelessWidget {
+  final String _title;
+  final String _body;
+  static const double _hPad = 16.0;
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/text_section.dart:7]: What is _hPad?
Use ALL_CAPS for constants.

File: security_drone_user_app/lib/presentation/text_section.dart [hunk 27]
```diff
@@ -0,0 +1,28 @@
...
+
+  TextSection(this._title, this._body);
+
+  @override
+  Widget build(BuildContext context) {
+    return  Column(
```
Comment (TomerMe2) [inline security_drone_user_app/lib/presentation/text_section.dart:13]: Remove space after return.

File: security_drone_user_app/lib/style.dart [hunk 28]
```diff
@@ -0,0 +1,28 @@
+import 'package:flutter/material.dart';
+
+const LargeTextSize = 26.0;
```
Comment (TomerMe2) [inline security_drone_user_app/lib/style.dart:3]: Use ALL_CAPS for constants

File: security_drone_user_app/test/thief_bloc_test.dart [hunk 29]
```diff
@@ -0,0 +1,40 @@
...
+  final file = new File(testsOutputFile);
+  file.writeAsStringSync('');
+}
+
+void main() {
+
+  var point = LatLngPoint(35.12, 34.11);
+  var entry = ThiefEntry(Image.asset("assets/images/thief.jpg"), DateTime.now(), point);
+
+  // say the we are testing
```
Comment (TomerMe2) [inline security_drone_user_app/test/thief_bloc_test.dart:19]: change the to that

File: security_drone_user_app/test/thief_bloc_test.dart [hunk 30]
```diff
@@ -0,0 +1,40 @@
...
+  debugTestingProd = 1;
+  var index = 0;
+
+  blocTest('Check clicked entry',
+      build: () => ThiefPageBloc.test(entry),
+      act: (bloc) => bloc.add(ThiefEntryClicked(index)),
+      expect: [ShowThiefEntry([entry],index)]
```
Comment (TomerMe2) [inline security_drone_user_app/test/thief_bloc_test.dart:26]: Add space after comma

**Review by TomerMe2 — CHANGES_REQUESTED** [review]
Nice work!
Please provide pictures of the screens that you've made.

File: security_drone_user_app/lib/data/models/dashboard_entry.dart [hunk 31]
```diff
@@ -0,0 +1,62 @@
...
+  DroneActivity activity;
+  DateTime startTime;
+  DateTime endTime;
+  MissionResultType missionResult;
+  String startReason;
+  String abortReason;
+
+  DashBoardEntry(this.activity, this.startTime, this.endTime, this.missionResult, this.startReason, this.abortReason);
+
+  static DashBoardEntry dummyFetch(){
```
[the remark this answers is not in the document]
Comment (shaharewq0) [reply security_drone_user_app/lib/data/models/dashboard_entry.dart:16]: תומר זאת קצת קטנוניות, האנרגיה שלי ושלך מבוזבזת על שטויות כאלה

File: security_drone_user_app/lib/data/models/dashboard_entry.dart [hunk 32]
```diff
@@ -0,0 +1,62 @@
...
+    return [
+      DashBoardEntry(droneActivity, DateTime.now(), DateTime.parse("1999-01-12"), MissionResultType.success, "trigger", ""),
+      DashBoardEntry(droneActivity, DateTime.now(), DateTime.parse("1999-01-12"), MissionResultType.success, "trigger", ""),
+      DashBoardEntry(droneActivity, DateTime.now(), DateTime.parse("1999-01-12"), MissionResultType.fail, "trigger", ""),
+      DashBoardEntry(droneActivity, DateTime.now(), DateTime.parse("1999-01-12"), MissionResultType.fail, "trigger", ""),
+    ];
+  }
+
+  @override
+  String toString() {
```
[the remark this answers is not in the document]
Comment (shaharewq0) [reply security_drone_user_app/lib/data/models/dashboard_entry.dart:41]: כרגע אין בכוונתי להתעסק בעיצוב האפלקציה ולכן מימשתי פונקציה מתאימה להצגת המידע הנדרש

File: security_drone_user_app/lib/presentation/text_section.dart (hunk 26 above)
Comment (shaharewq0) [reply security_drone_user_app/lib/presentation/text_section.dart:7]: לא הסטנדרט של DART שראיתי באינטרנט ובמדריכים
אני מבקש לרגיע קצת את הערות על סטייל הקוד. תתרכז בתוכן

File: security_drone_user_app/test/thief_bloc_test.dart (hunk 29 above)
Comment (shaharewq0) [reply security_drone_user_app/test/thief_bloc_test.dart:19]: לא מובנת לי הערה

Comment (TomerMe2) [reply security_drone_user_app/test/thief_bloc_test.dart:19]: the comment says "say the we are testing",
you should write instead "say that we are testing"

File: security_drone_user_app/lib/presentation/text_section.dart (hunk 26 above)
Comment (TomerMe2) [reply security_drone_user_app/lib/presentation/text_section.dart:7]: _hPad is not "style". It's content. Variables names are super important. If I cannot understand what _hPad means (and I can't), I will not be able to continue your code.
About the constant and final names, you are correct. While I learned Dart & Flutter, they used ALL_CPAS. They have changed the style recently. You can find a reference here: https://dart.dev/guides/language/effective-dart/style
Sorry about that, I will remove these comments.

File: security_drone_user_app/lib/data/models/thief_entry.dart [hunk 33]
```diff
@@ -0,0 +1,44 @@
...
+      ThiefEntry(Image.asset('assets/images/thief.jpg',fit: BoxFit.contain), DateTime.parse("1999-01-12"), LatLngPoint(12.0, 12.0)),
+      ThiefEntry(Image.asset('assets/images/thief.jpg',fit: BoxFit.contain), DateTime.parse("1999-01-12"), LatLngPoint(12.0, 12.0)),
+      ThiefEntry(Image.asset('assets/images/thief.jpg',fit: BoxFit.contain), DateTime.parse("1999-01-12"), LatLngPoint(12.0, 12.0))
+    ];
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (other is ThiefEntry){
+      return other.date == this.date && other.waypoint == this.waypoint;
```
[the remark this answers is not in the document]
Comment (shaharewq0) [reply security_drone_user_app/lib/data/models/thief_entry.dart:35]: solved using DateTime.compareTo

File: security_drone_user_app/lib/logic/dashboard_bloc.dart [hunk 34]
```diff
@@ -0,0 +1,37 @@
...
+    else if (event is RefreshDashboardEntries) {
+      print('api request to refresh the entries');
+      List<DashBoardEntry> entries = DashBoardEntry.dummyFetchAll();
+      yield ShowDashboardEntries(entries);
+    }
+    else if (event is DisplayDashboardEntries) {
+      yield ShowDashboardEntries(state.entries);
+    }
+    else {
+      throw Exception("Event not found $event");
```
[the remark this answers is not in the document]
Comment (shaharewq0) [reply security_drone_user_app/lib/logic/dashboard_bloc.dart:30]: we talked about this, an unrecognized event is a exception worthy error

File: security_drone_user_app/lib/main.dart (hunk 3 above)
Comment (shaharewq0) [reply security_drone_user_app/lib/main.dart:76]: to be dealt with in future issues

File: security_drone_user_app/lib/presentation/image_banner.dart (hunk 4 above)
Comment (shaharewq0) [reply security_drone_user_app/lib/presentation/image_banner.dart:14]: why?

File: security_drone_user_app/lib/presentation/pages/dashboard_page.dart (hunk 9 above)
Comment (shaharewq0) [reply security_drone_user_app/lib/presentation/pages/dashboard_page.dart:118]: only temporary

File: security_drone_user_app/lib/presentation/pages/patrol_set_page.dart (hunk 12 above)
Comment (shaharewq0) [reply security_drone_user_app/lib/presentation/pages/patrol_set_page.dart:34]: Saw a warning, it was suggested to implement the change.

**Comment by shaharewq0** [comment]
requested images:
Screenshot_1621086245
Screenshot_1621086256
Screenshot_1621086260
Screenshot_1621086273
Screenshot_1621086281
Screenshot_1621086220

**Comment by shaharewq0** [comment]
closes #19

**Comment by shaharewq0** [comment]
closes #16

**Comment by shaharewq0** [comment]
closes #21
